pinot-tools/src/main/java/com/linkedin/pinot/tools/ClusterStateVerifier.java

```diff
@@ -74,18 +77,42 @@ public Boolean call() {
    * @param tableNames list of table names which are about to scan
    * @return
    */
-  private boolean isClusterStable(final List<String> tableNames) {
-    LOGGER.info("Start scanning the stability of all the tables... Num of tables: {}", tableNames.size());
-    int numUnstablePartitions = 0;
-    for (String tableName : tableNames) {
-      numUnstablePartitions = super.isStable(tableName);
-      if (numUnstablePartitions != 0) {
-        LOGGER.error("Attention: Table {} is not stable. numUnstablePartitions: {} ", tableName, numUnstablePartitions);
-        break;
+  private boolean waitForClusterStable(final List<String> tableNames, final long timeoutSec)
+      throws InterruptedException{
+    final long startTimeMillis = System.currentTimeMillis();
+    final long maxEndTimeMillis = startTimeMillis + TimeUnit.MILLISECONDS.convert(timeoutSec, TimeUnit.SECONDS);
+
+    boolean stable = false;
+    int iteration = 0;
+    long sleepTimeMillis = (maxEndTimeMillis - startTimeMillis)/10;
+    if (sleepTimeMillis < MIN_SLEEP_BETWEEN_CHECKS_MILLIS) {
+      sleepTimeMillis = MIN_SLEEP_BETWEEN_CHECKS_MILLIS;
+    } else if (sleepTimeMillis > MAX_SLEEP_BETWEEN_CHECKS_MILLIS) {
+      sleepTimeMillis = MAX_SLEEP_BETWEEN_CHECKS_MILLIS;
+    }
+
+    while (!stable) {
+      iteration++;
+      LOGGER.info("Start scanning the stability of {} tables, iteration {}",
+          tableNames.size(), iteration);
+      int numUnstablePartitions = 0;
+      for (String tableName : tableNames) {
+        numUnstablePartitions = super.isStable(tableName);
+        if (numUnstablePartitions != 0) {
+          LOGGER.error("Table {} is not stable. numUnstablePartitions: {} ", tableName, numUnstablePartitions);
+          break;
+        }
+      }
+      stable = (numUnstablePartitions == 0);
+      if (!stable) {
+        if (System.currentTimeMillis() >= maxEndTimeMillis) {
```

Comment:
How important is it to account for the time to check stability ? I guess not much...but a simpler modification to the condition can be (current_time + sleepTime)  >= maxEndTime

Reply:
It is ok in this case. The time taken to check should be very small for a reasonable number of tables in the cluster.